Add ADR about using Vertex AI search (#124)
alphagov/govuk-chat: docs/adr/0006-vertex-ai-search.md

Reviewer: You could update the filename to match the title?

# 6. Suitability of using Search API V2 for retrieval

Date: 2025-03-24

## Context

> The Search API V2[1] app uses Google Vertex AI Search[2] (VAIS) to index and retrieve GOV.UK content, whereas GOV.UK Chat uses an OpenSearch cluster. Both applications operate on the same content (i.e. published documents) and provide results based on a search query.s

Reviewer: ```suggestion
The Search API V2[1] app uses Google Vertex AI Search[2] (VAIS) to index and retrieve GOV.UK content, whereas GOV.UK Chat uses an OpenSearch cluster. Both applications operate on the same content (i.e. published documents) and provide results based on a search query.
```

### Rationale
- VAIS has its own concept of chunking which it can do automatically. However this is an opaque system and it has few configuration options. The way we chunk content on GOV.UK Chat results in semantic chunks for a document, which vary in size depending on the content in that part of the page - this allows us to link to specific parts of a page on GOV.UK when we return sources to the user. VAIS chunking is instead only configurable at a size level so we can have either have only very small chunks or very large ones.
- VAIS does offer a service to "bring your own chunks", i.e. you chunk your content and then import those chunks instead of letting VAIS do it. A downside to this is that at the time of writing you can't import these chunks via an API like full document ingesting. You need to write the chunks to Google Cloud Storage as a JSONL file and then import them via an API call. This isn't impossible to do by any means, but it's a lot more work to store, process and clean up the data.
- With the "bring your own chunks" approach, you can't currently assign any metadata to the chunks. So we'd lose the ability to link to specific parts of a page on GOV.UK again.

> - The Search API V2 datastore is not chunked and were it be a chunked one this would incur a performance penalty for Search API V2 operations[3]. Therefore we believe that in order for Search API V2 to support providing chunked GOV.UK content this would need a secondary data store and decisions for any interactions with a data store as to whether it should apply to one or both data stores.

Reviewer: ```suggestion
- The Search API V2 datastore is not chunked and were it be a chunked one this would incur a performance penalty for Search API V2 operations[3]. Therefore we believe that in order for Search API V2 to support providing chunked GOV.UK content this would need a secondary data store, which would incur an additional decision point whenever interactions with datastore are considered as to whether they should apply to all data stores.
```

Tried to improve this a bit, not sure I've succeeded.

We do consider that should we enter a phase of work exploring the iteration of the retrieval logic of GOV.UK Chat then VAIS remains a platform that should be evaluated as a potential technology choice.

## Status
Accepted

## Consequences
- GOV.UK Chat will continue to have the responsibility of indexing GOV.UK content and will need to continue using an embedded model
- Search API V2 is not currently planned to provide an interface to access chunked GOV.UK Content and should other government teams need this information we would still need to work out the appropriate service for it
- GOV.UK Chat will not be tied to the roadmap of Search API V2 and could choose to explore other search technologies if and when retrieval is optimised
- GOV.UK Chat will continue to have the financial overhead of OpenSearch, which at low usage is substantially more expensive than VAIS
- GOV.UK Chat may need to consider offering an API for search results, or splitting search into a separate project should there be an established need for accessing chunked GOV.UK content by other teams or projects.

> [1] https://github.com/alphagov/search-api-v2
> [2] https://cloud.google.com/enterprise-search?hl=en
> [3] https://cloud.google.com/generative-ai-app-builder/docs/parse-chunk-documents#limitations
> [4] https://github.com/alphagov/search-api-v2/blob/1c1d9f05e2def51b035b19caa92cbcde55346e1c/app/models/concerns/publishing_api/content.rb#L9-L52

Reviewer: ```suggestion
[1]: https://github.com/alphagov/search-api-v2
[2]: https://cloud.google.com/enterprise-search?hl=en
[3]: https://cloud.google.com/generative-ai-app-builder/docs/parse-chunk-documents#limitations
[4]: https://github.com/alphagov/search-api-v2/blob/1c1d9f05e2def51b035b19caa92cbcde55346e1c/app/models/concerns/publishing_api/content.rb#L9-L52
```

These links are currently not rendering correctly in the preview view.
